src/analyze/gui/parser.cpp

```diff
@@ -670,9 +705,20 @@ void Parser::parseImpl(const QString& path, const QString& diffBase, const Filte
             }
             data->diff(diffData);
             data->stringCache.diffMode = true;
-        } else if (!data->read(stdPath, isReparsing)) {
-            emit failedToOpen(path);
-            return;
+        } else {
+            auto start = steady_clock::now();
```

Comment:
please don't open a thread just for progress reporting, instead add a callback to the parser interface and call it regularly (e.g. every 1000 lines or similar).

Reply:
I think you should move the progress reporting into the existing handleTimeStamp function and then emit the signal from there

Reply:
will do. I'll just have to figure out where to put the timer.

I might could put it in the `ParserState` though the logic could get a bit hairy given that a variable number of passes can happen.

Reply:
I also noticed:
 * `handleTimeStamp` only gets called on the first pass. I could change that logic, but am worried that has negative consequences.
 * `handleTimeStamp` doesn't currently have access to the parser which has the signals for `progress*` stuff. Not sure if passing that access in is the right thing to do, or to create new signals.

Reply:
Suggestion:

- take a progress callback (`std::function` e.g.) as argument in `ParserData()` ctor and store it in a member var
- also store the QElapsedTimer there, and start it from `Parser::parseImpl`
- always call handleTimeStamp and pass the `pass` as an argument (no pun intended)
- then call the callback and return early after it was called for anything but the first pass
- in `heaptrack_print.cpp` just return early if the pass isn't the first pass
- then you can pass a lambda to the `ParserData` from within `Parser::parseImpl` to do the progress tracking as you see fit. This should also make it hopefully easy to do proper progress reporting for the diff case, where we want to aggregate the two runs somehow

Reply:
I'll try to get to this tonight (US East Coast).

Reply:
Take a look at my latest commits that refactor according to the above spec and see how you feel. 

After looking at the complexity of my refactor, It occurs to me that the code with the new thread might be simpler, and was symmetric to the code above it, which itself opens an async job/thread for reading the diffData.

If my refactor attempt ended up too complex, I'm happy to change back to `std::async` based on your assessment.

Reply:
Hmm I seem to have broken the charts in the effort to refactor to the suggestion. Investigating...